File: doc-examples/example-kotlin/src/main/kotlin/example/PointSerde.kt
package example

import io.micronaut.core.type.Argument
import io.micronaut.serde.*
import jakarta.inject.Singleton

@Singleton // <1>
class PointSerde : Serde<Point> { // <2>
    override fun deserialize(
        decoder: Decoder,
        decoderContext: Deserializer.DecoderContext,
        type: Argument<in Point>
    ): Point {
        decoder.decodeArray().use { // <3>
            val x = it.decodeInt()
            val y = it.decodeInt()
            return Point.valueOf(x, y) // <4>
        }
    }

    override fun serialize(
        encoder: Encoder,
        context: Serializer.EncoderContext,
        value: Point,
        type: Argument<out Point>
    ) {
        val coords = value.coords()
        encoder.encodeArray(type).use { // <6>
            it.encodeInt(coords[0])
            it.encodeInt(coords[1])
        }
    }
}

File: doc-examples/example-python/src/main/python/example/PointSerde.py
from jakarta.inject import Singleton
from micronaut.core.type import Argument
from micronaut.serde import Decoder, Encoder, Serde
from micronaut.serde.Deserializer import DecoderContext
from micronaut.serde.Serializer import EncoderContext

from example.Point import Point


@Singleton  # <1>
class PointSerde(Serde[Point]):  # <2>
    def deserialize(self, decoder: Decoder, context: DecoderContext, type: Argument) -> Point:
        array = decoder.decodeArray()  # <3>
        x = array.decodeInt()
        y = array.decodeInt()
        array.finishStructure()
        return Point.value_of(x, y)  # <4>

    def serialize(self, encoder: Encoder, context: EncoderContext, type: Argument, value: Point) -> None:
        if value is None:
            raise ValueError("Point cannot be null")  # <5>
        coords = value.coords()
        array = encoder.encodeArray(type)  # <6>
        array.encodeInt(coords[0])
        array.encodeInt(coords[1])
        array.finishStructure()
